plugins/Vectorscope/VectorView.cpp

```diff
@@ -0,0 +1,284 @@
+	).count();
+	const unsigned int elapsed = currentTimestamp - m_persistTimestamp;
+	const unsigned int threshold = hq ? 10 : 50;
+	if (elapsed > threshold)
+	{
+		m_persistTimestamp = currentTimestamp;
+		// Non-HQ mode uses half the resolution → use limited buffer space.
+		const std::size_t useableBuffer = hq ? m_displayBuffer.size() : m_displayBuffer.size() / 4;
+		// The knob value is interpreted on log. scale, otherwise the effect would ramp up too slowly.
+		// Persistence value specifies fraction of light intensity that remains after 10 ms.
+		// → Compensate it based on elapsed time (exponential decay).
+		const float persist = log10(1 + 9 * m_controls->m_persistenceModel.value());
+		const float persistPerFrame = pow(persist, elapsed / 10.f);
+		// Note that for simplicity and performance reasons, this implementation only dims all stored
+		// values by a given factor. A true simulation would also do the inverse of desaturation that
+		// occurs in high-intensity traces in HQ mode.
+		for (std::size_t i = 0; i < useableBuffer; i++)
+		{
+			m_displayBuffer.data()[i] = m_displayBuffer.data()[i] * persistPerFrame;
+		}
+	}
+
+	// Get new samples from the lockless input FIFO buffer
+	auto inBuffer = m_bufferReader.read_max(m_inputBuffer->capacity());
+	std::size_t frameCount = inBuffer.size();
+
+	// Draw new points on top
+	float left, right;
+	int x, y;
+
+	const bool logScale = m_controls->m_logarithmicModel.value();
+	const unsigned short activeSize = hq ? m_displaySize : m_displaySize / 2;
+
+	// Helper lambda functions for better readability
+	// Make sure pixel stays within display bounds:
+	auto saturate = [=](short pixelPos) {return fmax(fmin(pixelPos, activeSize - 1), 0);};
+	// Take existing pixel and brigthen it. Very bright light should reduce saturation and become
+	// white. This effect is easily approximated by capping elementary colors to 255 individually.
+	auto updatePixel = [&](unsigned int x, unsigned int y, QColor addedColor)
+	{
+		QColor currentColor = ((QRgb*)m_displayBuffer.data())[x + y * activeSize];
+		currentColor.setRed(fmin(currentColor.red() + addedColor.red(), 255));
+		currentColor.setGreen(fmin(currentColor.green() + addedColor.green(), 255));
+		currentColor.setBlue(fmin(currentColor.blue() + addedColor.blue(), 255));
+		((QRgb*)m_displayBuffer.data())[x + y * activeSize] = currentColor.rgb();
+	};
+
+	if (hq)
+	{
+		// High quality mode: check distance between points and draw a line.
+		// The longer the line is, the dimmer, simulating real electron trace on luminescent screen.
+		for (std::size_t frame = 0; frame < frameCount; frame++)
+		{
+			// Scale left and right channel from (-1.0, 1.0) to display range
+			if (logScale)
+			{
+				// To better preserve shapes, the log scale is applied to the distance from origin,
+				// not the individual channels.
+				const float distance = sqrt(pow(inBuffer[frame][0], 2) + pow(inBuffer[frame][1], 2));
```

Comment:
It seems like manual multiplication is faster than using `pow()` in some cases. It's not the case when using GCC with `-ffast-math`, but I don't know about other compilers.
See: https://baptiste-wicht.com/posts/2017/09/cpp11-performance-tip-when-to-use-std-pow.html

Reply:
You are right; I just checked at https://gcc.godbolt.org/ and while trying a few compilers none of them ended up even remotely close to the few instructions that a simple multiplication produced.

Reply:
Correction: it seems that with -O3 the results are generally much better. Sometimes even exactly the same. I tried to see what would happen if I used a `square` lambda that implemented the multiplication, but in general it seems to make things worse. So a simple `num * num` would be probably the safest bet for performance (even though it makes the line a bit too long).

Reply:
How about `hypot`?
https://en.cppreference.com/w/cpp/numeric/math/hypot

Reply:
Hard to say. GCC apparently produces just 8 instructions with -O3 for manual `sqrt(num*num)` and a function call to who knows what when `hypot` is used (insides of the function were not shown). So if anything, it is probably slightly worse performance-wise.
MSVC produced a call either way (to sqrt or hypot) so the basic overhead is the same and then it would depend on what exactly is inside. Overall it is probably safer to just use `sqrt(num*num)`, although the code does not look as nice..

Reply:
It is not `sqrt(num*num)`, it is `sqrt(num1*num1 + num2*num2)`.

`hypot(num1, num2)` beats `sqrt(num1*num1 + num2*num2)` on all optimization levels on GCC.

https://godbolt.org/z/wZdQ2-

Reply:
Ouch, damn, I mixed up the code with my previous experiments. Thanks, I'll look at it once more.

Reply:
Hm, I still cannot confirm it. Each implementation still has a function call (either for sqrt or for hypot) and since they are inside a library, I can't get it to show the instructions.

Reply:
I decided to leave it as it is; if anything, google searches turn out questions about why is `hypot` _slower_ than doing the same thing by hand, and after implementing the zoom everything fits into one line again, so no issue with readability either.